onnx/optimizer/optimize.h

```diff
@@ -32,9 +32,10 @@ struct Optimizer {
     passes[sp->name] = std::move(sp);
   }
 
-  std::unique_ptr<onnx::ModelProto> optimize(std::unique_ptr<onnx::ModelProto> mp_in, std::vector<std::string>& names) {
-
-    std::shared_ptr<onnx::Graph> g(onnx::ImportModelProto(*mp_in));
+  onnx::ModelProto optimize(
```

Comment:
I'm hesitating.... The current way of using unique_ptr will make peak memory usage less, although it enforces callers to follow the programming API somehow.

Reply:
Just curious, where does `unique_ptr` help reduce the peak memory usage?

Reply:
@linkerzhang It shouldn't. If the caller side doesn't need the original model anymore, it can move when passing the model proto into the optimizer.

Reply:
Yep. "it can move when passing the model proto into the optimizer." this is the usage of unique_ptr, I think :). Double check, does ModelProto have move constructor please? I thought it didn't have.

Reply:
No you missed my point. I mean the optimizer doesn't care about the ownership of the model proto that is passed in. It's should be the caller of the optimizer to decide whether the original model is still needed. If you use unique_ptr here, you are forcing the caller to either give up the ownership or make another useless copy.

Reply:
Hmmm, so the caller code may be,
 ModelProto modelProto = ... ; // load model from a file.
 Optimizer optimizer;
 optimizer.optimize(modelProto, {"pass1", "pass2"});

The caller of optimizer has no good way to release the memory before optimizer to lower the peak memory usage, am I right? however, if there's a move constructor of ModelProto (needs double check), then it's still doable for caller of optimizer to write code like, optimizer.optimize(std::move(modelProto), {"pass1"});

This is not a strong point for now as we may only have offline tools calling optimizer codes (in that case, peak memory is probably not a concern)

Reply:
The peak memory usage actually matters since I have seen OOM in onnx-caffe2's CI :-)
And I don't think this change will increase any memory footprint. Using your example:
```
ModelProto modelProto = ...;
```
If we use unique_ptr, then the user would need to make a second copy and wrap that with a unique_ptr to pass into the optimizer, it has two copies during the whole time optimizer is running. While with my change, no matter whether ModelProto has move ctor or not, since we are passing model proto as const ref, it only has one copy, so it actually uses less memory.

Reply:
if using unique_ptr, then the code will look like,

std::unique_ptr<ModelProto> modelProto(new ModelProto);
....

peak memory usage should be lower down as there's always one ModelProto "data".

Reply:
@linkerzhang Sorry I really don't get how after this change it can become worse. Yes if the caller already has a unique_ptr, both interfaces will only have one copy (note the new interface is taking const reference, so it doesn't need the caller to make another copy). But if what the user has is an object, then using unique_ptr will cause a second copy while the new interface still only need one copy.

Reply:
OK, my bad that I didn't clarify clearly. You're right. With current implementation, there's no difference.

I'm thinking of whether we should really do in-place change. Like you mentioned, having a unique_ptr<ModelProto> sent from caller, means the caller gives the ownership to us, we could own it and change it in-place accordingly. Our current implementation is copying all data into graph object, right? (say, weights). This part could be avoided. By using const ModelProto&, our implementation inside will have to make a copy since it's not clear that we could own the ModelProto object sent in.